io-engine/src/bdev/nexus/nexus_bdev_children.rs

```diff
@@ -815,37 +816,70 @@ impl<'n> Nexus<'n> {
+            );
+
+            return None;
+        };
+
+        if Ok(ChildState::Open)
+            == c.state
+                .compare_exchange(ChildState::Open, ChildState::Faulted(reason))
         {
-            if Ok(ChildState::Open)
-                == c.state.compare_exchange(
-                    ChildState::Open,
-                    ChildState::Faulted(reason),
-                )
-            {
-                warn!("{:?}: faulted with {}, will retire", c, reason);
-
-                // The child state was not faulted yet, so this is the first
-                // I/O to this child for which we
-                // encountered an error.
-                Reactors::master().send_future(Nexus::child_retire_routine(
-                    name,
-                    child_device.to_owned(),
-                    retry,
-                ));
-            } else {
-                warn!("{:?}: faulted with {}, already retired", c, reason);
-            }
+            warn!("{c:?}: faulted with {reason}, will retire");
+
+            // The child state was not faulted yet, so this is the first
+            // I/O to this child for which we
+            // encountered an error.
+            Reactors::master().send_future(Nexus::child_retire_routine(
+                name,
+                child_device.to_owned(),
+                retry,
+            ));
         } else {
-            error!(
-                "{:?}: trying to retire a child device which \
-                        cannot be not found '{}'",
-                self, child_device
+            warn!("{c:?}: faulted with {reason}, already retired");
+        }
+
+        // Get the existing rebuild log, or create a new one.
+        let log = c.get_or_init_rebuild_log();
+
+        // Attach the rebuild log all the channels. It is done asynchronously
+        // via `traverse_io_channels`. That implies that other I/O requests
+        // may arrive to other channel(s) in parallel before the log
+        // is attached to that channel(s).
+        //
+        // If that happens and a write I/O to another channel succeeds
+        // (e.g. the condition causing I/O failure is erratic),
+        // the data goes through to the device and there's no problem
+        // if this write operation is not logged.
+        //
+        // If the parallel write operation fails, `retire_child_device` is
+        // called anyway, and that operation ends up logged.
+        // However, in this case, log attach will be called again,
+        // but that's okay since the channel checks if the same log is attached.
+        //
+        // Having a critical section covering log attachment would make the code
+        // validity easier to reason about, but that isn't practical as
+        // it would require a lock in I/O code path, and a complicated
+        // syncronization with SPDK-driven `traverse_io_channels`.
+        if let Some(ref log) = log {
```

Comment:
What if the retire happens before we run this piece of code, won't we miss IOs ?

Reply:
Yes indeed. Thanks  for pointing out. Will fix.